pallets/file-system/src/utils.rs

```diff
@@ -176,10 +203,104 @@ where
+        who: T::AccountId,
+        location: FileLocation<T>,
+        root: FileKey<T>,
+        proof: Proof<T>,
+    ) -> DispatchResult {
+        let bsp =
+            <T::Providers as storage_hub_traits::ReadProvidersInterface>::get_provider(who.clone())
+                .ok_or(Error::<T>::NotAProvider)?;
+
+        // Check that the provider is indeed a BSP.
+        ensure!(
+            <T::Providers as storage_hub_traits::ReadProvidersInterface>::is_bsp(&bsp),
+            Error::<T>::NotABsp
+        );
+
+        // Check that the storage request exists.
+        let file_metadata =
+            <StorageRequests<T>>::get(&location).ok_or(Error::<T>::StorageRequestNotFound)?;
+
+        expect_or_err!(
+                    file_metadata.bsps_confirmed < file_metadata.bsps_required,
+                    "Storage request should never have confirmed bsps equal to or greater than required bsps, since they are deleted when it is reached.",
+                    Error::<T>::StorageRequestBspsRequiredFulfilled,
+                    bool
+                );
+
+        // Check that the BSP has volunteered for the storage request.
+        ensure!(
+            <StorageRequestBsps<T>>::contains_key(&location, &who),
+            Error::<T>::BspNotVolunteered
+        );
+
+        // Check that the storage provider has not already confirmed storing the file.
+        ensure!(
+            !<StorageRequestBsps<T>>::get(&location, &who)
+                .expect("BSP should exist since we checked it above")
+                .confirmed,
+            Error::<T>::BspAlreadyConfirmed
+        );
+
+        let mut file_metadata =
+            <StorageRequests<T>>::get(&location).ok_or(Error::<T>::StorageRequestNotFound)?;
+
+        // Check that the number of confirmed bsps is less than the required bsps and increment it.
+        match file_metadata
+            .bsps_confirmed
+            .checked_add(&T::StorageRequestBspsRequiredType::one())
+        {
+            Some(bsps_confirmed) => {
+                file_metadata.bsps_confirmed = bsps_confirmed;
+            }
+            None => {
+                return Err(Error::<T>::StorageRequestBspsRequiredFulfilled.into());
+            }
+        }
+
+        // Check that the proof is valid.
+        <T::ProofDealer as storage_hub_traits::ProofsDealerInterface>::verify_proof(
+            &bsp, &root, &proof,
+        )?;
+
+        // Remove storage request if we reached the required number of bsps.
+        // TODO this should be modified to also take into account if the MSP confirmed storing the file once we add MSP functionality.
+        if file_metadata.bsps_confirmed == file_metadata.bsps_required {
+            // Remove storage request metadata.
+            <StorageRequests<T>>::remove(&location);
+
+            const REMOVE_LIMIT: u32 = u32::MAX;
+
+            // Remove storage request bsps
+            let _ = <StorageRequestBsps<T>>::clear_prefix(&location, REMOVE_LIMIT, None);
```

Comment:
We should handle this properly, and set a proper `REMOVE_LIMIT`, to avoid attacks because this function takes too long. `REMOVE_LIMIT` should be equal to the number of BSPs who volunteered for this file. We can keep a counter in the storage request for that if needed. And after setting the `REMOVE_LIMIT` to that value, if the returned value of `clear_prefix` is something other than `None`, then that's a big unexpected error. As it would mean that there is an inconsistency between the BSPs accounted for as volunteers, and the number of keys in `StorageRequestBsps` for that `location`